api/v1alpha1/httproute_types.go

```diff
@@ -34,11 +34,23 @@ type HTTPRouteSpec struct {
 
 // HTTPRouteHost is the configuration for a given host.
 type HTTPRouteHost struct {
-	// Hostnames are a list of hosts names that match this host
-	// block.
+	// Hostname is the fully qualified domain name of a network host,
+	// as defined by RFC 3986. Note the following deviations from the
+	// "host" part of the URI as defined in the RFC:
 	//
-	// TODO: RFC link
-	Hostnames []string `json:"hostnames"`
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//
+	// Incoming requests are matched against Hostname before processing HTTPRoute
+	// rules. If Hostname is unspecified, the Gateway routes all traffic based on
+	// the specified rules.
```

Comment:
Does `Hostname` need to be unique among `HTTPRoute` resources?

How does SNI fit in?  I would guess that for a TLS connection, first some certificate must match, and then the ingress controller uses the first route that also matches?

Reply:
Why do we need to singularize the hostname?
There’s a use case for having multiple hostnames in one endpoint, for example multi language site through example.TLD.
A client can select to one or the other hostname and the developer wants not to duplicate her `Rules`. If we singularize we need to duplicate the `Rules`.

Reply:
I think this should be singular but no unique.
Some use-cases create multiple routes with the same hosts but different paths (or other HTTP metadata). And these routes can belong to different teams. 

> Translating the above doc to ingress v2, I would expect the following flow for ingress https connections:

TLS and Routing details should not be clubbed together. Although, there is leakage between the layers of the OSI model, they are largely different layers. @danehans, in your description, I think Step 2 and 3 should be changed. `HTTPRoute` can't be resolved until after TLS is done. One can use only the TLS config setting of a route, but shouldn't select that route. Route selection is based on other metadata in the HTTP reqeust.


Regarding, Sandor's point, I think this can be opened up to allow for prefix and suffix wildcards in future if there is a need for it.

Reply:
In the case of HTTPRoutes, the gateway can terminate TLS and then use HTTP to select a HTTPRoute, but in the case of a TCPRoute, HTTP headers are unavailable.  So faced with an incoming TLS request, the gateway could first use SNI to build a set of candidate routes, which must comprise either multiple HTTPRoutes or a single TCPRoute, and then the gateway could either (a) terminate TLS and use HTTP metadata to select a single HTTPRoute or (b) use the TLSRoute.

Reply:
> I think Step 2 and 3 should be changed

@haiyanmeng I submitted https://github.com/kubernetes-sigs/service-apis/pull/68 to doc a typical ingress request flow, ptal.

Reply:
Reference for uniqueness: https://github.com/kubernetes-sigs/service-apis/issues/67

Reply:
I'm inclined to merge this and continue conversation about the wildcard on an issue specific to having wildcards.

Reply:
I extracted the use case that @szuecs raised into #80.